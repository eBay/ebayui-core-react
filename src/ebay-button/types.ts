export type Priority = 'primary' | 'secondary' | 'tertiary' | 'none'

export type Variant = 'standard' | 'destructive' | 'form'

export type Size = 'default' /* DEPRECATED */ | 'regular' | 'large'

export type BodyState = 'loading' | 'expand' | 'reset' | 'none'

export type Split = 'start' | 'end';
