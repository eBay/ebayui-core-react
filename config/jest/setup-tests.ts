import '@testing-library/jest-dom'

declare var global: typeof globalThis;

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: () => {} });
}
