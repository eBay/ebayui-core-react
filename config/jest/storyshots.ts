import { render } from "@testing-library/react";
import initStoryshotsAddon from "@storybook/addon-storyshots";
import type { StoryshotsOptions } from "@storybook/addon-storyshots/dist/ts3.9/api/StoryshotsOptions";

const reactTestingLibrarySerializer = {
    print: (val, serialize, indent) => serialize(val.asFragment()),
    test: (val) => val && val.hasOwnProperty("container"),
};

export function initStoryshots(config: StoryshotsOptions) {
    initStoryshotsAddon({
        ...config,
        renderer: render,
        snapshotSerializers: [reactTestingLibrarySerializer],
    });
}
