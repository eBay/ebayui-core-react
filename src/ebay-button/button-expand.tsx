import React, { FC } from 'react'
import EbayButtonCell from './button-cell'
import EbayButtonText from './button-text'
import { EbayIcon } from '../ebay-icon'

const EbayButtonExpand: FC = ({ children }) =>
    children ? (
        <EbayButtonCell>
            <EbayButtonText>
                {children}
            </EbayButtonText>
            <EbayIcon name="dropdown" />
        </EbayButtonCell>
    ) : (
        <EbayIcon name="dropdown" />
    )

export default EbayButtonExpand
