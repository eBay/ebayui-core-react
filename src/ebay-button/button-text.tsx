import React, { ComponentProps, FC } from 'react'
import classNames from 'classnames'

type Props = ComponentProps<'span'>

const EbayButtonText: FC<Props> = ({ className, children }) => (
    <span className={classNames(className, 'btn__text')}>{children}</span>
)

export default EbayButtonText
