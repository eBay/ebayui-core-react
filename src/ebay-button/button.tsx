import React, {
    Children,
    ComponentProps,
    FC,
    KeyboardEvent,
    MouseEvent,
    ReactElement,
    ReactNode,
    RefObject
} from 'react'
import classNames from 'classnames'
import { withForwardRef } from '../common/component-utils'
import { Priority, Size, BodyState, Variant, Split } from './types'
import { EbayIcon } from '../ebay-icon'
import EbayButtonLoading from './button-loading'
import EbayButtonExpand from './button-expand'

export type EbayButtonProps = {
    fluid?: boolean;
    partiallyDisabled?: boolean;
    truncate?: boolean;
    href?: string;
    priority?: Priority;
    variant?: Variant;
    size?: Size;
    bodyState?: BodyState,
    split?: Split;
    transparent?: boolean;
    onClick?: (e: MouseEvent) => void;
    onEscape?: (e: KeyboardEvent<HTMLButtonElement | HTMLAnchorElement>) => void;
    forwardedRef?: RefObject<HTMLAnchorElement & HTMLButtonElement>;
    borderless?: boolean;
    fixedHeight?: boolean;
}

type Props = ComponentProps<'button'> & ComponentProps<'a'> & EbayButtonProps;

const EbayButton:FC<Props> = ({
    priority = 'secondary',
    variant = 'standard',
    size = 'regular',
    bodyState,
    split,
    transparent = false,
    fluid = false,
    disabled,
    partiallyDisabled,
    children,
    onEscape = () => {},
    truncate = false,
    href,
    className: extraClasses,
    forwardedRef,
    borderless,
    fixedHeight,
    ...rest
}) => {
    const classPrefix = href ? 'fake-btn' : 'btn'
    const priorityStyles: { [key in Priority]: string } = {
        primary: `${classPrefix}--primary`,
        secondary: `${classPrefix}--secondary`,
        tertiary: `${classPrefix}--tertiary`,
        none: ''
    }
    const sizeStyles: { [key in Size]: string } = {
        large: `${classPrefix}--large`,
        regular: '',
        default: ''
    }
    const splitStyles: { [key in Split]: string } = {
        start: `${classPrefix}--split-start`,
        end: `${classPrefix}--split-end`
    }
    const isDestructive = variant === 'destructive'
    const isForm = variant === 'form'
    const isLoading = bodyState === 'loading'
    const isExpand = bodyState === 'expand'
    const isSlim = isForm && (isIconOnly(children) || (isExpand && !children))
    const className = classNames(
        classPrefix,
        extraClasses,
        priorityStyles[isForm || borderless ? 'none' : priority],
        sizeStyles[size],
        splitStyles[split],
        isDestructive && `${classPrefix}--destructive`,
        isForm && `${classPrefix}--form`,
        isSlim && `${classPrefix}--slim`,
        transparent && `${classPrefix}--transparent`,
        fluid && `${classPrefix}--fluid`,
        truncate && `${classPrefix}--truncated`,
        borderless && `${classPrefix}--borderless`,
        fixedHeight && (sizeStyles[size] ? `${sizeStyles[size]}-${fixedHeight}` : `${classPrefix}--fixed-height`)
    )

    const onKeyDown = (e: KeyboardEvent<HTMLButtonElement | HTMLAnchorElement>) => {
        if (e.key === 'Escape' || e.key === 'Esc') {
            onEscape(e)
        }
    }

    const bodyContent = getBodyContent(children, { isLoading, isExpand })
    const ariaLive = isLoading ? `polite` : null

    return href ? (
        <a
            className={className}
            href={disabled ? undefined : href}
            ref={forwardedRef}
            onKeyDown={onKeyDown}
            aria-live={ariaLive}
            {...rest}
        >
            {bodyContent}
        </a>
    ) : (
        <button
            disabled={disabled}
            aria-disabled={partiallyDisabled}
            aria-live={ariaLive}
            className={className}
            ref={forwardedRef}
            onKeyDown={onKeyDown}
            {...rest}
        >
            {bodyContent}
        </button>
    )
}

type bodyContentOptions = {
    isLoading: boolean;
    isExpand: boolean;
}

function getBodyContent(children:ReactNode, { isLoading, isExpand }: bodyContentOptions) {
    switch (true) {
        case isLoading:
            return <EbayButtonLoading />
        case isExpand:
            return <EbayButtonExpand>{children}</EbayButtonExpand>
        default:
            return children
    }
}

function isIconOnly(children: ReactNode): boolean {
    const childrenArray = Children.toArray(children) as ReactElement[]
    return childrenArray.length === 1 && childrenArray[0].type === EbayIcon
}

export default withForwardRef(EbayButton)
