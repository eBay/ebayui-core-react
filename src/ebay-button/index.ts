export { default as EbayButton, type EbayButtonProps } from './button'
export { default as EbayButtonCell } from './button-cell'
export type { Size, Priority } from './types'
