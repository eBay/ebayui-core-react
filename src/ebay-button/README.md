# EbayButton

## Demo
[Storybook](https://opensource.ebay.com/ebayui-core-react/main/?path=/story/ebay-button--default)

## Usage

### Import JS
```jsx harmony
import { EbayButton } from '@ebay/ui-core-react/ebay-button'
```
### Import following styles from SKIN
```jsx harmony
import "@ebay/skin/button"
```
### Or import styles using SCSS/CSS
```jsx harmony
import '@ebay/skin/button.css'
```

### Big button
```jsx harmony
<EbayButton size="large">I'm a big button!</EbayButton>
```

### Icon button
```jsx
<EbayButton aria-label="Menu button">
    <EbayIcon name="MENU" />
</EbayButton>
```

### Complex button
```jsx
<EbayButton priority="primary" fluid>
    <EbayButtonCell style={{ justifyContent: 'space-between' }}>
        <span>Label</span>
        <span>Filter</span>
    </EbayButtonCell>
</EbayButton>
```

## Attributes

Name | Type | Stateful | Required | Description | Data
--- | --- | --- | --- | --- | ---
`priority` | String | No | No | `primary`, `secondary` (default), `tertiary`, `none`
`variant` | String | Yes | No | `standard` (default), `destructive`, `form`
`size` | String | No | No | `regular` (default), `large`
`bodyState` | String | No | No | `loading`, `expand`
`href` | String | No | No | for link that looks like a button
`fluid` | Boolean | No | No | takes the whole width of the parent element
`disabled` | Boolean | Yes | No
`partiallyDisabled` | Boolean | No | No | sets `aria-disabled` but not `disabled` prop
`transparent` | Boolean | Yes | No | transparent background color (overrides `priority` prop)
`truncate` | Boolean | No | No | will truncate the text of the button onto a single line, and adds an ellipsis, when the button's text overflows
`borderless` | Boolean | No | No | shows button without border
`fixedHeight` | Boolean | No | No | fixes the height based on `size`
`onClick` | Function | - | No | click or action key pressed (`Space` / `Enter`)
`onEscape` | Function | - | No | `Esc`-key pressed
`onFocus` | Function | - | No | triggered on focus
`onBlur` | Function | - | No | triggered on blur
