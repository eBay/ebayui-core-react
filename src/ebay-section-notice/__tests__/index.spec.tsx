import React from 'react';
import { render } from '@testing-library/react';
import { initStoryshots } from '../../../config/jest/storyshots';
import { EbayButton } from '../../ebay-button'
import { EbayNoticeContent, EbaySectionNotice } from '../index';

describe('<EbaySectionNotice>', () => {
    describe('when a button is added in the children of the main section notice', () => {
        it('should render the button as well as the notice content', () => {
            const wrapper = render(
                <EbaySectionNotice>
                    <EbayNoticeContent>
                        <p id="content">Here is some content</p>
                    </EbayNoticeContent>
                    <EbayButton id="actionButton">Action</EbayButton>
                </EbaySectionNotice>
            )

            expect(wrapper.getByText('Here is some content')).toBeTruthy()
            expect(wrapper.getByRole('button').textContent).toContain('Action')
        })
    })

    describe('when aria-label is given', () => {
        it('should set SVG role=`img`', () => {
            const wrapper = render(
                <EbaySectionNotice status="confirmation" aria-label="Success Confirmation">
                    <EbayNoticeContent>
                        <p>You did it.</p>
                    </EbayNoticeContent>
                </EbaySectionNotice>
            )
            expect(wrapper.container.querySelector('svg')).toHaveAttribute('role', 'img')
        })

        it('should add the text to the main section as an aria-label if there is no status', () => {
            const wrapper = render(
                <EbaySectionNotice aria-label="Success Confirmation">
                    <EbayNoticeContent>
                        <p>You did it.</p>
                    </EbayNoticeContent>
                </EbaySectionNotice>
            )

            expect(wrapper.getByRole('region')).toHaveAttribute('aria-label', 'Success Confirmation')
        })
    })

    describe('on using the notice with no content', () => {
        it('should throw an error', () => {
            jest.spyOn(console, 'error').mockImplementation(() => null)
            expect(() => {
                render(<EbaySectionNotice />)
            }).toThrowError(`EbaySectionNotice: Please use a EbayNoticeContent that defines the content of the notice`);
            (console.error as jest.Mock).mockRestore()
        })
    })
})

initStoryshots({
    config: ({ configure }) =>
        configure(() => {
            require('./index.stories')
        }, module)
})
