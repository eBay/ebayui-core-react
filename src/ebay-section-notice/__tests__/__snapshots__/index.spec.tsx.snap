// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-section-notice Attention message 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="section-notice-attention"
    aria-roledescription="Notice"
    class="section-notice section-notice--attention"
    role="region"
  >
    <div
      class="section-notice__header"
      id="section-notice-attention"
    >
      <svg
        aria-hidden="true"
        class="icon icon--attention-filled-small"
        focusable="false"
        height="16px"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-attention-filled-small"
        />
      </svg>
    </div>
    <div
      class="section-notice__main"
    >
      <p>
        <strong>
          Error.
        </strong>
         Please take another look at the following:
        <br />
        <a
          href="http://www.ebay.com"
        >
          Card number
        </a>
        ,
        <a
          href="http://www.ebay.com"
        >
          Expiration date
        </a>
         &
        <a
          href="http://www.ebay.com"
        >
          Security code
        </a>
        .
      </p>
    </div>
    <div
      class="section-notice__footer"
    >
      <button
        class="btn fake-link btn--secondary"
      >
        Show more
      </button>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-section-notice Confirmation message 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="section-notice-confirmation"
    aria-roledescription="Notice"
    class="section-notice section-notice--confirmation"
    role="region"
  >
    <div
      class="section-notice__header"
      id="section-notice-confirmation"
    >
      <svg
        aria-hidden="true"
        class="icon icon--confirmation-filled-small"
        focusable="false"
        height="16px"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-confirmation-filled-small"
        />
      </svg>
    </div>
    <div
      class="section-notice__main"
    >
      <p>
        This successfully finished 
        <a
          href="http://www.ebay.com"
        >
          next page
        </a>
      </p>
    </div>
    <div
      class="section-notice__footer"
    >
      <button
        class="btn fake-link btn--secondary"
      >
        Dismiss
      </button>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-section-notice Default message (with action) 1`] = `
<DocumentFragment>
  <section
    aria-roledescription="Notice"
    class="section-notice"
    role="region"
  >
    <div
      class="section-notice__main"
    >
      <p>
        Items you didn't win will now show in the 
        <a
          href="http://www.ebay.com"
        >
          Didn't win
        </a>
         section of this page.
      </p>
    </div>
    <div
      class="section-notice__footer"
    >
      <button
        class="btn fake-link btn--secondary"
      >
        Do something
      </button>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-section-notice Default message (with no action) 1`] = `
<DocumentFragment>
  <section
    aria-roledescription="Notice"
    class="section-notice"
    role="region"
  >
    <div
      class="section-notice__main"
    >
      <p>
        Items you didn't win will now show in the 
        <a
          href="http://www.ebay.com"
        >
          Didn't win
        </a>
         section of this page.
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-section-notice Information message 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="section-notice-information"
    aria-roledescription="Notice"
    class="section-notice section-notice--information"
    role="region"
  >
    <div
      class="section-notice__header"
      id="section-notice-information"
    >
      <svg
        aria-hidden="true"
        class="icon icon--information-filled-small"
        focusable="false"
        height="16px"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-information-filled-small"
        />
      </svg>
    </div>
    <div
      class="section-notice__main"
    >
      <p>
        <strong>
          Good news!
        </strong>
         You get free shipping on your next pair of shoes!Â 
        <a
          href="http://www.ebay.com"
        >
          Learn more
        </a>
        .
      </p>
    </div>
    <div
      class="section-notice__footer"
    >
      <button
        class="btn fake-link btn--secondary"
      >
        Dismiss
      </button>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-section-notice Section with link 1`] = `
<DocumentFragment>
  <section
    aria-roledescription="Notice"
    class="section-notice"
    role="region"
  >
    <div
      class="section-notice__main"
    >
      <h2
        class="section-notice__title"
      >
        Title
      </h2>
      <p>
        Items you didn't win will now show in the 
        <a
          href="http://www.ebay.com"
        >
          Didn't win
        </a>
         section of this page.
      </p>
    </div>
    <div
      class="section-notice__footer"
    >
      <a
        href="https://www.ebay.com"
      >
        Go see details
      </a>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-section-notice Section with title 1`] = `
<DocumentFragment>
  <section
    aria-roledescription="Notice"
    class="section-notice"
    role="region"
  >
    <div
      class="section-notice__main"
    >
      <h2
        class="section-notice__title"
      >
        Title
      </h2>
      <p>
        Items you didn't win will now show in the 
        <a
          href="http://www.ebay.com"
        >
          Didn't win
        </a>
         section of this page.
      </p>
    </div>
  </section>
</DocumentFragment>
`;
